contracts/programs/data-feeds-cache/src/lib.rs

```diff
@@ -0,0 +1,951 @@
+        Discriminator,
+    };
+    use state::{ReceivedDecimalReport, WorkflowMetadata};
+
+    use crate::{
+        common::ANCHOR_DISCRIMINATOR,
+        event::LegacyFeedsReported,
+        state::{CacheTransmission, DecimalReport, LegacyFeedEntry, WritePermissionFlag},
+    };
+
+    use super::*;
+
+    // todo, add feed admins? who will also be a zero_copy array
+    pub fn initialize(ctx: Context<Initialize>, feed_admins: Vec<Pubkey>) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_init()?;
+        state.owner = ctx.accounts.owner.key();
+
+        let mut prev_admin = Pubkey::default();
+        for admin in feed_admins.iter() {
+            require!(
+                &prev_admin <= admin,
+                DataCacheError::AddressesMustStrictlyIncrease
+            );
+            state.feed_admins.push(admin.clone());
+            prev_admin = *admin;
+        }
+
+        // todo: can probably remove
+        state.proposed_owner = Pubkey::default();
+
+        let (_, authority_nonce) = Pubkey::find_program_address(
+            &[b"legacy_writer", ctx.accounts.state.key().as_ref()],
+            &crate::ID,
+        );
+        state.legacy_writer_nonce = authority_nonce;
+
+        Ok(())
+    }
+
+    pub fn set_feed_admin(ctx: Context<SetFeedAdmin>, admin: Pubkey, is_admin: bool) -> Result<()> {
+        let mut state = ctx.accounts.state.load_mut()?;
+
+        match state.feed_admins.binary_search_by(|a| a.cmp(&admin)) {
+            Ok(i) => {
+                if !is_admin {
+                    state.feed_admins.remove(i);
+                }
+            }
+            Err(i) => {
+                if is_admin {
+                    state.feed_admins.insert(i, admin);
+                }
+            }
+        };
+        Ok(())
+    }
+
+    pub fn transfer_ownership(
+        ctx: Context<TransferOwnership>,
+        proposed_owner: Pubkey,
+    ) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_mut()?;
+        state.proposed_owner = proposed_owner;
+
+        Ok(())
+    }
+
+    pub fn accept_ownership(ctx: Context<AcceptOwnership>) -> Result<()> {
+        let state = &mut ctx.accounts.state.load_mut()?;
+
+        state.owner = state.proposed_owner;
+        state.proposed_owner = Pubkey::default();
+
+        Ok(())
+    }
+
+    pub fn init_decimal_reports<'info>(
+        ctx: Context<'_, '_, 'info, 'info, InitDecimalReports<'info>>,
+        data_ids: Vec<[u8; 16]>,
+    ) -> Result<()> {
+        // check feed admin here
+        let state = &mut ctx.accounts.state.load()?;
+        verify_feed_admin(&ctx.accounts.feed_admin, &state.feed_admins)?;
+
+        let state_key = ctx.accounts.state.key();
+
+        let data_ids_account_infos = ctx.remaining_accounts;
+
+        require!(
+            data_ids.len() == data_ids_account_infos.len(),
+            DataCacheError::ArrayLengthMismatch
+        );
```

Comment:
nit:(not a big deal, but auditors tend to nitpick this)
```suggestion
        require_eq!(
            data_ids.len(), data_ids_account_infos.len(),
            DataCacheError::ArrayLengthMismatch
        );
```